notebooks/02_model/cornac_bpr_deep_dive.ipynb

```diff
@@ -0,0 +1,601 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<i>Copyright (c) Microsoft Corporation. All rights reserved.</i>\n",
+    "\n",
+    "<i>Licensed under the MIT License.</i>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Bayesian Personalized Ranking (BPR)\n",
+    "\n",
+    "This notebook serves as an introduction to Bayesian Personalized Ranking (BPR) model for implicit feedback.  In this tutorial, we focus on learning the BPR model using matrix factorization approach, hence, the model is sometimes also named BPRMF.\n",
+    "\n",
+    "The implementation of the model is from [Cornac](https://github.com/PreferredAI/cornac), which is a framework for recommender systems with a focus on models leveraging auxiliary data (e.g., item descriptive text and image, social network, etc)."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 0 Global Settings and Imports"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "System version: 3.6.8 |Anaconda, Inc.| (default, Feb 21 2019, 18:30:04) [MSC v.1916 64 bit (AMD64)]\n",
+      "Cornac version: 1.1.2\n"
+     ]
+    }
+   ],
+   "source": [
+    "import sys\n",
+    "sys.path.append(\"../../\")\n",
+    "import time\n",
+    "import os\n",
+    "import cornac\n",
+    "import papermill as pm\n",
+    "import pandas as pd\n",
+    "from reco_utils.dataset import movielens\n",
+    "from reco_utils.dataset.python_splitters import python_random_split\n",
+    "from reco_utils.evaluation.python_evaluation import (map_at_k, ndcg_at_k, precision_at_k, recall_at_k)\n",
```

Comment:
here, unless we add a new line, the parenthesis are not needed

Reply:
updated!